spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -858,6 +858,23 @@ public static String quotedFullIdentifier(String catalogName, Identifier identif
         .quoted();
   }
 
+  public static org.apache.spark.sql.execution.datasources.PartitionSpec getInferredSpec(
+      SparkSession spark, Path rootPath) {
+    FileStatusCache fileStatusCache = FileStatusCache.getOrCreate(spark);
+    InMemoryFileIndex fileIndex =
+        new InMemoryFileIndex(
```

Comment:
Add files will do this InMemoryFileIndex calculation twice (first it will do here to infer the spec with option.empty()) and later it will be called in getPartitions (see L902) to list the files. I assume since fileStatusCache is passed with same spark session it will not do recomputation?

Else I will calculate the InMemoryFileIndex in AddFilesProcedure and infer schema there, and pass it along to getPartitions()

Reply:
... I think this is ok .. I worry a little bit about making caches that we never clear but I don't think we have a safe way of invalidating this cache either.